src/tracing/trace_flamegraph.c

```diff
@@ -0,0 +1,312 @@
+	char *used_fname;
+
+	tmp_flamegraph_context = xdmalloc(sizeof(xdebug_trace_flamegraph_context));
+	tmp_flamegraph_context->trace_file = xdebug_trace_open_file(fname, script_filename, options, (char**) &used_fname);
+	tmp_flamegraph_context->trace_filename = used_fname;
+
+	return tmp_flamegraph_context->trace_file ? tmp_flamegraph_context : NULL;
+}
+
+void *xdebug_trace_flamegraph_init_cost(char *fname, zend_string *script_filename, long options)
+{
+	xdebug_trace_flamegraph_context *tmp_flamegraph_context = xdebug_trace_flamegraph_init(fname, script_filename, options);
+
+	if (tmp_flamegraph_context) {
+		tmp_flamegraph_context->mode = XDEBUG_TRACE_OPTION_FLAMEGRAPH_COST;
+		tmp_flamegraph_context->functions = xdebug_hash_alloc(64, (xdebug_hash_dtor_t) fg_function_dtor);
+	}
+
+	return tmp_flamegraph_context;
+}
+
+void *xdebug_trace_flamegraph_init_mem(char *fname, zend_string *script_filename, long options)
+{
+	xdebug_trace_flamegraph_context *tmp_flamegraph_context = xdebug_trace_flamegraph_init(fname, script_filename, options);
+
+	if (tmp_flamegraph_context) {
+		tmp_flamegraph_context->mode = XDEBUG_TRACE_OPTION_FLAMEGRAPH_MEM;
+		tmp_flamegraph_context->functions = xdebug_hash_alloc(64, (xdebug_hash_dtor_t) fg_function_dtor);
+	}
+
+	return tmp_flamegraph_context;
+}
+
+void xdebug_trace_flamegraph_deinit(void *ctxt)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+
+	if (NULL == context) {
+		return;
+	}
+
+	if (NULL != context->functions) {
+		xdebug_hash_destroy(context->functions);
+		context->functions = NULL;
+	}
+
+	if (NULL != context->trace_file) {
+		fclose(context->trace_file);
+		context->trace_file = NULL;
+		xdfree(context->trace_filename);
+		context->trace_filename = NULL;
+	}
+
+	xdfree(context);
+}
+
+void xdebug_trace_flamegraph_write_header(void *ctxt)
+{
+}
+
+void xdebug_trace_flamegraph_write_footer(void *ctxt)
+{
+}
+
+char *xdebug_trace_flamegraph_get_filename(void *ctxt)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+
+	return context->trace_filename;
+}
+
+void xdebug_trace_flamegraph_function_entry(void *ctxt, function_stack_entry *fse, int function_nr)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+	function_stack_entry            *parent_fse;
+	flamegraph_function             *function;
+	flamegraph_function             *parent_function;
+	xdebug_str                      *prefix = xdebug_str_new();
+	char                            *tmp_name;
+
+	tmp_name = xdebug_show_fname(fse->function, 0, 0);
+	function = fg_function_ctor();
+
+	parent_fse = fg_parent_find();
+
+	if (NULL == parent_fse) {
+		/* No parent means we are top-level, prefix is function name. */
+		xdebug_str_add_fmt(prefix, tmp_name);
+	} else {
+		/* Find value in our custom hashmap in order to compute prefix. */
+		parent_function = fg_function_find(context, parent_fse->function_nr);
```

Comment:
I still don't understand and/or agree with using a separate hashmap for this. This information can easily be part of the `function_stack_entry`, just like Xdebug does for the profiler: 

Definition at https://github.com/xdebug/xdebug/blob/master/src/lib/lib.h#L130-L136
Part of FSE at: https://github.com/xdebug/xdebug/blob/master/src/lib/lib.h#L171

There would be no need for looking up hashes either, as the parent prefix can be found in `parent_fse`.

Reply:
I don't like too but I was unsure if you'd allow me to change the fse struct, if you're OK with adding info in this struct, I'd gladly revert to something simpler !

Reply:
Yeah, that's fine... although I've been thinking is whether it could just reuse the "profiler" part of it, perhaps with some changes?

Reply:
I'm actually on hollydays but I'll continue this when I'll get back at work, in 10 days ish.